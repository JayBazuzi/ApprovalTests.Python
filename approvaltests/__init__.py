"""Approvaltests module."""
from .version import *
from .utilities import *
from .approvals import *
from .asserts import *

__author__ = "Chris Lucian; Llewellyn Falco; Jim Counts; Tim Preece"
__version__ = version_number
