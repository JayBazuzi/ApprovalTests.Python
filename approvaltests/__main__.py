from .verify import verify_using_commandline_arguments

verify_using_commandline_arguments()