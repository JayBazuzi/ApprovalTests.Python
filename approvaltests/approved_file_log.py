from pathlib import Path

APPROVAL_TESTS_TEMP_DIRECTORY: str = ".approval_tests_temp"


class ApprovedFilesLog:
    @staticmethod
    def clear_log_file() -> None:
        ApprovedFilesLog.get_approved_files_log().write_text("")

    @staticmethod
    def get_approved_files_log() -> Path:
        path = Path("%s/.approved_files.log" % APPROVAL_TESTS_TEMP_DIRECTORY)
        path.parent.mkdir(parents=True, exist_ok=True)

        return path

    @staticmethod
    def log(approved_file):
        with ApprovedFilesLog.get_approved_files_log().open(mode="a") as file:
            file.write(f"{approved_file}\n")


ApprovedFilesLog.clear_log_file()
