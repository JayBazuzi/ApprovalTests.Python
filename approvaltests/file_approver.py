import filecmp
import os
import pathlib
from typing import Optional, Callable

from approvaltests.core.comparator import Comparator
from approvaltests.core.namer import Namer
from approvaltests.core.reporter import Reporter
from approvaltests.core.writer import Writer


def exists(path: str) -> bool:
    return os.path.isfile(path)


class ReporterNotWorkingException(Exception):
    def __init__(self, reporter: Reporter):
        super().__init__(f"Reporter {reporter} failed to work!")


class FileComparator(Comparator):
    def compare(self, received_path: str, approved_path: str) -> bool:
        if not exists(approved_path) or not exists(received_path):
            return False
        if filecmp.cmp(approved_path, received_path):
            return True
        try:
            approved_raw = pathlib.Path(approved_path).read_text()
            approved_text = approved_raw.replace("\r\n", "\n")
            received_raw = pathlib.Path(received_path).read_text()
            received_text = received_raw.replace("\r\n", "\n")
            return approved_text == received_text
        except BaseException:
            return False


class FileApprover:
    previous_approved = []
    allowed_duplicates = []

    @staticmethod
    def verify(
        namer: Namer, writer: Writer, reporter: Reporter, comparator: Comparator
    ) -> Optional[str]:
        approved = namer.get_approved_filename()
        received = namer.get_received_filename()

        if (FileApprover.is_this_a_multiple_verify(approved) ):
            return (
                f"We noticed that you called verify more than once in the same test. Is that what you want to do?\n"
                f"\tApproved file name is: {approved}\n"
            )
        FileApprover.previous_approved.append(approved)

        # The writer has the ability to change the name of the received file
        received = writer.write_received_file(received)
        verified = FileApprover.verify_files(approved, received, reporter, comparator)

        if not verified:
            return (
                f"Approval Mismatch, received != approved\n"
                f"\tApproved: {approved}\n"
                f"\tReceived: {received} "
            )
        return None

    @staticmethod
    def is_this_a_multiple_verify(approved):
        return approved in FileApprover.previous_approved \
            and not FileApprover.is_duplicate_allowed(approved)

    @staticmethod
    def is_duplicate_allowed(approved):
        for allowed in FileApprover.allowed_duplicates:
            if allowed(approved):
                return True
        return False

    @staticmethod
    def verify_files(
        approved_file: str,
        received_file: str,
        reporter: Reporter,
        comparator: Comparator,
    ) -> bool:
        if comparator.compare(received_file, approved_file):
            os.remove(received_file)
            return True

        worked = reporter.report(received_file, approved_file)
        if not worked:
            raise ReporterNotWorkingException(reporter)
        return False

    @staticmethod
    def add_allowed_duplicates(is_duplicate_allowed: Callable[[str],bool]):
        FileApprover.allowed_duplicates.append(is_duplicate_allowed)
