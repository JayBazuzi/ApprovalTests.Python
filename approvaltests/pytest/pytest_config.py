class PytestConfig:
    test_naming_pattern = "test_*"

    @staticmethod
    def set_config(config):
        PytestConfig.test_naming_pattern = config.getini("python_functions")


def pytest_config(config):
    PytestConfig.set_config(config)
