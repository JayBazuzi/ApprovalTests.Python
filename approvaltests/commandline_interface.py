from pathlib import Path

from approvaltests import verify
from build.lib.approvaltests.commandline_interface import parse_arguments
from namer.cli_namer import CliNamer


# simplify the following lines of code from commandline_interface.py

# 1.  Modify so our future-selves would recognize it
# >>> 2.  Make a test
# 3.  Refactor parse_arguments to be part of a class (suggested by Jay)

# NEw code...
# received = args.received or stdin.read()
# Old code...
# received = args.received
# if args.received is None:
#    received = stdin.read()
# return (args.id, received)



def verify_using_commandline_arguments():
    test_id, received = parse_arguments()
    verify_with_id(received, test_id)


def verify_with_id(received, test_id):
    verify(received, namer=CliNamer(test_id=test_id))
    print(f"Test Passed: {test_id}")


if __name__ == "__main__":
    verify_using_commandline_arguments()


def list_approved_files_in_test_directory(test_dir: Path):
    return [file.name for file in test_dir.iterdir() if file.name.endswith(".approved.txt")]
