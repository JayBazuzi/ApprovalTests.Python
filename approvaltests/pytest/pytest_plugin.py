from approvaltests import pytest_config

def pytest_configure(config):
    pytest_config(config)
