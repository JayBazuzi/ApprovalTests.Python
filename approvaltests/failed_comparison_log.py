from pathlib import Path

from approvaltests.approved_file_log import APPROVAL_TESTS_TEMP_DIRECTORY


class FailedComparisonLog:
    @staticmethod
    def clear_log_file() -> None:
        FailedComparisonLog.get_failed_comparison_log().write_text("")

    @staticmethod
    def get_failed_comparison_log() -> Path:
        path = Path("%s/.failed_comparison.log" % APPROVAL_TESTS_TEMP_DIRECTORY)
        path.parent.mkdir(parents=True, exist_ok=True)

        return path

    @staticmethod
    def log(received_file, approved_file):
        with FailedComparisonLog.get_failed_comparison_log().open(mode="a") as file:
            file.write(f"{received_file} -> {approved_file}\n")


FailedComparisonLog.clear_log_file()
