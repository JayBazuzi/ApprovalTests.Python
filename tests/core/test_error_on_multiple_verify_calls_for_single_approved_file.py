import pytest

from approvaltests import verify, ApprovalException, settings, approvals, Options
from approvaltests.file_approver import FileApprover
from tests.namers.test_templated_namer import TemplatedCustomNamer


def test_multiple_calls_to_verify():
    verify("call to verify")
    interceeding_verify()
    with pytest.raises(ApprovalException):
        verify("call to verify")


def interceeding_verify():
    namer = TemplatedCustomNamer(
        "{test_source_directory}/differenttest.{approved_or_received}.txt"
    )
    verify("# call to verify", options=Options().with_namer(namer))


def test_allow_multiple_verifies_per_method():
    settings().allow_multiple_verify_calls_for_this_method()
    verify("call to verify")
    verify("call to verify")


def test_allow_multiple_verifies_by_lambda():
    filename = approvals.get_default_namer().get_approved_filename()
    FileApprover.add_allowed_duplicates(lambda n: n == filename)
    FileApprover.add_allowed_duplicates(lambda n: False)
    verify("call to verify")
    verify("call to verify")


def test_single_call_to_verify():
    verify("call to verify")


def test_help_message():
    verify(FileApprover.get_duplicate_verify_error_message("<approved_file>"))
