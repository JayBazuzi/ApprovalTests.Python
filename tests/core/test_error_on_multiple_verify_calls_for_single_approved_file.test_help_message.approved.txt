We noticed that you called verify more than once in the same test. 
This is the second call to verify:
    approved_file: <approved_file>

By default, ApprovalTests only allows one verify() call per test.
To find out more, visit: 
https://github.com/approvals/ApprovalTests.Python/blob/main/docs/how_to/multiple_approvals_per_test.md

# Possible Fixes
1. Separate your test into two tests
2. In your verify call, add `options=NamerFactory.with_parameters("your_paramater")`
3. In your test, call `approvals.settings().allow_multiple_verify_calls_for_this_method()`
